File: chrome/android/java/src/org/chromium/chrome/browser/UmaBridge.java
// Copyright 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser;

/**
 * Static methods to record user actions.
 *
 * We have a different native method for each action as we want to use c++ code to
 * extract user command keys from code and show them in the dashboard.
 * See: chromium/src/content/browser/user_metrics.h for more details.
 */
public class UmaBridge {

    /**
     * Record that the user opened the menu.
     */
    public static void menuShow() {
        nativeRecordMenuShow();
    }

    /**
     * Record that the user opened the menu through software menu button.
     * @param isByHwButton
     * @param isDragging
     */
    public static void usingMenu(boolean isByHwButton, boolean isDragging) {
        nativeRecordUsingMenu(isByHwButton, isDragging);
    }

    private static native void nativeRecordMenuShow();
    private static native void nativeRecordUsingMenu(boolean isByHwButton, boolean isDragging);
}

File: chrome/browser/android/metrics/uma_bridge.cc
// Copyright 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/android/metrics/uma_bridge.h"

#include <jni.h>

#include "base/metrics/histogram.h"
#include "content/public/browser/user_metrics.h"
#include "jni/UmaBridge_jni.h"

using base::UserMetricsAction;
using content::RecordAction;

static void RecordMenuShow(JNIEnv*, jclass) {
  RecordAction(UserMetricsAction("MobileMenuShow"));
}

static void RecordUsingMenu(JNIEnv*,
                            jclass,
                            jboolean is_by_hw_button,
                            jboolean is_dragging) {
  if (is_by_hw_button) {
    if (is_dragging) {
      NOTREACHED() << "We do not support dragging for hardware menu button.";
    } else {
      RecordAction(UserMetricsAction("MobileUsingMenuByHwButtonTap"));
    }
  } else {
    if (is_dragging) {
      RecordAction(UserMetricsAction("MobileUsingMenuBySwButtonDragging"));
    } else {
      RecordAction(UserMetricsAction("MobileUsingMenuBySwButtonTap"));
    }
  }
}

// Android Beam

static void RecordBeamCallbackSuccess(JNIEnv*, jclass) {
  RecordAction(UserMetricsAction("MobileBeamCallbackSuccess"));
}

static void RecordBeamInvalidAppState(JNIEnv*, jclass) {
  RecordAction(UserMetricsAction("MobileBeamInvalidAppState"));
}

// First Run Experience
static void RecordFreSignInShown(JNIEnv*, jclass) {
  RecordAction(UserMetricsAction("MobileFre.SignInShown"));
}

namespace chrome {
namespace android {

// Register native methods
bool RegisterUmaBridge(JNIEnv* env) {
  return RegisterNativesImpl(env);
}

}  // namespace android
}  // namespace chrome
